packages/core/src/global.js

```diff
@@ -38,111 +37,96 @@ export let Global: React.StatelessFunctionalComponent<
-        .nextElementSibling
-      this.sheet.before = ((element: any): Element | null)
-      this.sheet.flush()
-    }
-    this.props.cache.insert(``, this.props.serialized, this.sheet, false)
-  }
 
-  componentWillUnmount() {
-    this.sheet.flush()
+    return (
+      <style
+        {...{
+          [`data-emotion-${cache.key}`]: serializedNames,
+          dangerouslySetInnerHTML: { __html: rules },
+          nonce: cache.sheet.nonce
+        }}
+      />
+    )
   }
-  render() {
-    if (!isBrowser) {
-      let { serialized } = this.props
-
-      let serializedNames = serialized.name
-      let serializedStyles = serialized.styles
-      let next = serialized.next
-      while (next !== undefined) {
-        serializedNames += ' ' + next.name
-        serializedStyles += next.styles
-        next = next.next
-      }
-
-      let shouldCache = this.props.cache.compat === true
 
-      let rules = this.props.cache.insert(
-        ``,
-        { name: serializedNames, styles: serializedStyles },
-        this.sheet,
-        shouldCache
+  // yes, i know these hooks are used conditionally
+  // but it is based on a constant that will never change at runtime
+  // it's effectively like having two implementations and switching them out
+  // so it's not actually breaking anything
+
+  let sheetRef = React.useRef()
+
+  React.useLayoutEffect(
+    () => {
+      let sheet = new StyleSheet({
+        key: `${cache.key}-global`,
+        nonce: cache.sheet.nonce,
+        container: cache.sheet.container
+      })
+      // $FlowFixMe
+      let node: HTMLStyleElement | null = document.querySelector(
+        `style[data-emotion-${cache.key}="${serialized.name}"]`
       )
 
-      if (!shouldCache) {
-        return (
-          <style
-            {...{
-              [`data-emotion-${this.props.cache.key}`]: serializedNames,
-              dangerouslySetInnerHTML: { __html: rules },
-              nonce: this.props.cache.sheet.nonce
-            }}
-          />
-        )
+      if (node !== null) {
+        sheet.tags.push(node)
       }
-    }
-    return null
-  }
-}
+      if (cache.sheet.tags.length) {
+        sheet.before = cache.sheet.tags[0]
+      }
+      sheetRef.current = sheet
+      return () => {
+        sheet.flush()
+      }
+    },
+    [cache]
+  )
+
+  React.useLayoutEffect(
+    () => {
+      if (serialized.next !== undefined) {
+        // insert keyframes
+        insertStyles(cache, serialized.next, true)
+      }
+      let sheet: StyleSheet = ((sheetRef.current: any): StyleSheet)
+      if (sheet.tags.length) {
+        // if this doesn't exist then it will be null so the style element will be appended
+        let element = sheet.tags[sheet.tags.length - 1].nextElementSibling
+        sheet.before = ((element: any): Element | null)
+        sheet.flush()
+      }
+      cache.insert(``, serialized, sheet, false)
+    },
+    [cache, serialized]
```

Comment:
previous version was only updating styles on `serialized.name` change, I'm not sure if the previous version was safe - because I don't quite understand the cursor/keyframes thing and what guarantees it gives in regards to `serialized.name`. Would love to learn about it though, could you explain this a little bit? 

If we can't put `serialized.name` in the inputs array then we can remove this array altogether, as `serialized` changes with each render - so technically this input array doesn't give us anything right now

Reply:
If `serialized.name` is the same, the style is the same since the hash of the contents of the keyframes is in the style it's part of the whole styles' hash.